src/openrct2-ui/windows/PatrolArea.cpp

```diff
@@ -0,0 +1,304 @@
+            gMapSelectPositionA = posA;
+            gMapSelectPositionB = posB;
+            map_invalidate_selection_rect();
+        }
+    }
+
+    void OnToolAbort(rct_widgetindex widgetIndex) override
+    {
+        hide_gridlines();
+        ClearPatrolAreaToRender();
+        gfx_invalidate_screen();
+    }
+
+    void OnToolDown(rct_widgetindex widgetIndex, const ScreenCoordsXY& screenCoords) override
+    {
+        auto mapTile = GetBestCoordsFromPos(screenCoords);
+        if (mapTile)
+        {
+            auto staff = GetEntity<Staff>(_staffId);
+            if (staff != nullptr)
+            {
+                _mode = staff->IsPatrolAreaSet(*mapTile) ? StaffSetPatrolAreaMode::Unset : StaffSetPatrolAreaMode::Set;
+            }
+        }
+
+        OnToolDrag(widgetIndex, screenCoords);
+    }
+
+    void OnToolDrag(rct_widgetindex widgetIndex, const ScreenCoordsXY& screenCoords) override
+    {
+        auto staff = GetEntity<Staff>(_staffId);
+        if (staff != nullptr)
+        {
+            MapRange range(gMapSelectPositionA, gMapSelectPositionB);
+            auto staffSetPatrolAreaAction = StaffSetPatrolAreaAction(_staffId, range, _mode);
+            GameActions::Execute(&staffSetPatrolAreaAction);
+        }
+    }
+
+    EntityId GetStaffId() const
+    {
+        return _staffId;
+    }
+
+    void SetStaffId(EntityId staffId)
+    {
+        _staffId = staffId;
+        EnableTool();
+    }
+
+private:
+    EntityId _staffId;
+    StaffSetPatrolAreaMode _mode;
+
+    void EnableTool()
+    {
+        show_gridlines();
+        if (!tool_set(this, 0, Tool::WalkDown))
+        {
+            input_set_flag(INPUT_FLAG_6, true);
+            show_gridlines();
+            SetPatrolAreaToRender(_staffId);
+            gfx_invalidate_screen();
+        }
+    }
+
+    void InputSize()
+    {
+        Formatter ft;
+        ft.Add<int16_t>(MINIMUM_TOOL_SIZE);
+        ft.Add<int16_t>(MAXIMUM_TOOL_SIZE);
+        WindowTextInputOpen(this, WIDX_PREVIEW, STR_SELECTION_SIZE, STR_ENTER_SELECTION_SIZE, ft, STR_NONE, STR_NONE, 3);
+    }
+
+    bool PatrolAreaToolIsActive()
+    {
+        if (!(input_test_flag(INPUT_FLAG_TOOL_ACTIVE)))
+            return false;
+        if (gCurrentToolWidget.window_classification != WC_PATROL_AREA)
+            return false;
+        return true;
+    }
+
+    bool IsStaffWindowOpen()
+    {
+        // If staff window for this patrol area was closed, tool is no longer active
+        auto staffWindow = window_find_by_number(WC_PEEP, _staffId);
+        return staffWindow != nullptr;
+    }
+
+    std::optional<CoordsXY> GetBestCoordsFromPos(const ScreenCoordsXY& pos)
+    {
+        auto coords = footpath_get_coordinates_from_pos(pos, nullptr, nullptr);
+        return coords.IsNull() ? std::nullopt : std::make_optional(coords);
+    }
+};
+
+rct_window* WindowPatrolAreaOpen(EntityId staffId)
+{
+    auto current = reinterpret_cast<PatrolAreaWindow*>(window_find_by_class(WC_PATROL_AREA));
```

Comment:
i don't quite see the reason for this code. It basically does focus or create then does focus or create again. Did you test it without this bit of code?

Reply:
It needs to take into account the staff ID is the same. I can change `WindowFocusOrCreate` to just `WindowCreate` if that is less confusing.

Reply:
But don't you force the staff ID in the bit of code after the focus or create

Reply:
I don't know whether it is better to switch the staff ID or to re-create the window which will reset the position of it. What do you think?